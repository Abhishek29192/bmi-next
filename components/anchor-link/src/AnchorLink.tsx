import React, { useContext } from "react";
import classnames from "classnames";
import Icon from "@bmi/icon";
import { withClickable } from "@bmi/clickable";
import Link, { LinkProps } from "@material-ui/core/Link";
import { ColorPairContext } from "@bmi/color-pair";
import arrowForwardIcon from "./icons/arrow-right.svg";
import styles from "./AnchorLink.module.scss";

export type Props = Omit<LinkProps, "color"> & {
  children: React.ReactNode;
  isDisabled?: boolean;
  iconStart?: boolean;
  iconEnd?: boolean;
  iconInverted?: boolean;
  color?: "default" | "black" | "white";
};

const typeToColorMap: Record<"dark" | "light", Props["color"]> = {
  dark: "white",
  light: "black"
};

const AnchorLink = ({
  children,
  isDisabled,
  iconStart,
  iconEnd,
  className,
  color,
  iconInverted,
  ...rest
}: Props) => {
  const { type, theme } = useContext(ColorPairContext);
  const colorFromTheme: Props["color"] =
    color || (type && typeToColorMap[type]) || "default";

  const arrowIcon = (
    <Icon
      source={arrowForwardIcon}
      aria-hidden="true"
      focusable="false"
      className={classnames(styles["icon"], {
        [styles["icon--inverted"]!]: iconInverted
      })}
    />
  );

  return (
    <Link
      className={classnames(className, styles["Anchorlink"], {
        [styles["Anchorlink--disabled"]!]: isDisabled,
        [styles[`Anchorlink--${colorFromTheme}`]!]:
          theme !== "white" && colorFromTheme !== "default"
      })}
      {...rest}
    >
      {iconStart ? arrowIcon : null}
      <span>{children}</span>
      {iconEnd ? arrowIcon : null}
    </Link>
  );
};

export default withClickable(AnchorLink);
