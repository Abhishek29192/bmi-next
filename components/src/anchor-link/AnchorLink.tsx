import React, { useContext } from "react";
import { Link, LinkProps } from "@material-ui/core";
import classnames from "classnames";
import Icon from "../icon/Icon";
import { withClickable } from "../clickable/Clickable";
import { ColorPairContext } from "../color-pair/ColorPair";
import arrowForwardIcon from "./icons/arrow-right.svg";
import launchIcon from "./icons/launch.svg";
import styles from "./AnchorLink.module.scss";

export type Props = Omit<LinkProps, "color"> & {
  children: React.ReactNode;
  isDisabled?: boolean;
  iconStart?: boolean;
  iconEnd?: boolean;
  isExternal?: boolean;
  iconInverted?: boolean;
  color?: "default" | "black" | "white";
  action?: { href?: string };
};

const typeToColorMap: Record<"dark" | "light", Props["color"]> = {
  dark: "white",
  light: "black"
};

const AnchorLink = ({
  children,
  isDisabled,
  iconStart,
  iconEnd,
  className,
  isExternal,
  color,
  iconInverted,
  ...rest
}: Props) => {
  const { type, theme } = useContext(ColorPairContext);
  const colorFromTheme: Props["color"] =
    // eslint-disable-next-line security/detect-object-injection
    color || (type && typeToColorMap[type]) || "default";
  const icon = (
    <Icon
      data-testid={isExternal ? "external" : "internal"}
      source={isExternal ? launchIcon : arrowForwardIcon}
      aria-hidden="true"
      focusable="false"
      className={classnames(styles["icon"], {
        [styles["icon--inverted"]!]: iconInverted
      })}
    />
  );

  return (
    <Link
      className={classnames(className, styles["Anchorlink"], {
        [styles["Anchorlink--disabled"]!]: isDisabled,
        [styles[`Anchorlink--${colorFromTheme}`]!]:
          theme !== "white" && colorFromTheme !== "default"
      })}
      {...rest}
    >
      {iconStart ? icon : null}
      <span>{children}</span>
      {iconEnd ? icon : null}
    </Link>
  );
};

export default withClickable(AnchorLink);
