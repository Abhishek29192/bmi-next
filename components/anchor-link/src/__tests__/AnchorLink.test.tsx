import React from "react";
import AnchorLink from "..";
import { render } from "@testing-library/react";

describe("AnchorLink component", () => {
  it("renders default correctly", () => {
    const { container } = render(
      <AnchorLink href="/" to="/">
        BMI Group
      </AnchorLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders disabled correctly", () => {
    const { container } = render(
      <AnchorLink href="/" to="/" isDisabled>
        BMI Group
      </AnchorLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders leading icon correctly", () => {
    const { container } = render(
      <AnchorLink href="/" to="/" iconStart>
        BMI Group
      </AnchorLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders trailing icon correctly", () => {
    const { container } = render(
      <AnchorLink href="/" to="/" iconEnd>
        BMI Group
      </AnchorLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders icon disabled correctly", () => {
    const { container } = render(
      <AnchorLink href="/" to="/" iconEnd isDisabled>
        BMI Group
      </AnchorLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders multi-line with icon correctly", () => {
    const { container } = render(
      <AnchorLink href="/" to="/" iconEnd>
        <p style={{ width: "300px", textAlign: "left" }}>
          Sed ut perspiciatis unde omnis iste natus error sit voluptatem
          accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae
          ab illo inventore veritatis et quasi architecto beatae vitae dicta
          sunt explicabo.
        </p>
      </AnchorLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
