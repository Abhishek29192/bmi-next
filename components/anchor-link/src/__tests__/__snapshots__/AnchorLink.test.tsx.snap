// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnchorLink component renders default correctly 1`] = `
<a
  class="MuiTypography-root MuiLink-root MuiLink-underlineHover Anchorlink MuiTypography-colorPrimary"
  href="/"
  to="/"
>
  <span>
    BMI Group
  </span>
</a>
`;

exports[`AnchorLink component renders disabled correctly 1`] = `
<a
  class="MuiTypography-root MuiLink-root MuiLink-underlineHover Anchorlink Anchorlink--disabled MuiTypography-colorPrimary"
  href="/"
  to="/"
>
  <span>
    BMI Group
  </span>
</a>
`;

exports[`AnchorLink component renders icon disabled correctly 1`] = `
<a
  class="MuiTypography-root MuiLink-root MuiLink-underlineHover Anchorlink Anchorlink--disabled MuiTypography-colorPrimary"
  href="/"
  to="/"
>
  <span>
    BMI Group
  </span>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
    />
  </svg>
</a>
`;

exports[`AnchorLink component renders leading icon correctly 1`] = `
<a
  class="MuiTypography-root MuiLink-root MuiLink-underlineHover Anchorlink MuiTypography-colorPrimary"
  href="/"
  to="/"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
    />
  </svg>
  <span>
    BMI Group
  </span>
</a>
`;

exports[`AnchorLink component renders multi-line with icon correctly 1`] = `
<a
  class="MuiTypography-root MuiLink-root MuiLink-underlineHover Anchorlink MuiTypography-colorPrimary"
  href="/"
  to="/"
>
  <span>
    <p
      style="width: 300px; text-align: left;"
    >
      Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
    </p>
  </span>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
    />
  </svg>
</a>
`;

exports[`AnchorLink component renders trailing icon correctly 1`] = `
<a
  class="MuiTypography-root MuiLink-root MuiLink-underlineHover Anchorlink MuiTypography-colorPrimary"
  href="/"
  to="/"
>
  <span>
    BMI Group
  </span>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
    />
  </svg>
</a>
`;
